storage/client/client.h

```diff
@@ -0,0 +1,55 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_STORAGE_CLIENT_CLIENT_H_
+#define GOOGLE_CLOUD_CPP_STORAGE_CLIENT_CLIENT_H_
+
+#include "storage/client/bucket_metadata.h"
+#include "storage/client/credentials.h"
+#include "storage/client/status.h"
+
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+/**
+ * Define the interface used to communicate with Google Cloud Storage.
+ *
+ * This is a dependency injection point so higher-level abstractions (like
+ * `storage::Bucket` or `storage::Object`) can be effectively tested.
+ */
+class Client {
+ public:
+  virtual ~Client() = default;
+
+  // The member functions of this class are not intended for general use by
+  // application developers (they are simply a dependency injection point). Make
+  // them protected, so the mock classes can override them, and then make the
+  // classes that do use them friends.
+ protected:
+  friend class Bucket;
+  virtual std::pair<Status, BucketMetadata> BucketGet(
```

Comment:
Why `BucketGet` instead of `GetBucket`?

Reply:
Ah, because the JSON API is called `Bucket: get`:

https://cloud.google.com/storage/docs/json_api/v1/buckets/get

oops, actually Bucket**s**.  I am not sure if it is a good choice, happy to change it if there are any thoughts one what we should call these. @houglum should have a say too.

Reply:
I have no strong feelings on `BucketGet` vs. `GetBucket` vs. the plural versions of either of those, but please choose something that will be consistent, e.g.,

1. [verb]–[noun], e.g., `GetBucket`
2. matches the API param ordering, e.g., `BucketGet`
3. some other pattern?

I prefer (1) because that's how we typically write APIs, independent of their REST resource-ordering in URIs, but that's just me. Besides, if it's in alpha, you can change your mind on the APIs, right? 😄